OpenRA.Game/WorldUtils.cs

```diff
@@ -41,6 +41,113 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+			if (target.X < source.X)
+				xOver = -1024;
+			else
+				xOver = 1024;
+
+			if (target.Y < source.Y)
+				yOver = -1024;
+			else
+				yOver = 1024;
+
+			WVec dir = new WVec(xOver, yOver, 0);
+			var overselect = dir * (1 + (rayWidth.Length / 1024));
+			var finalTarget = target + overselect;
+			var finalSource = source - overselect;
+
+			var actorsInSquare = world.ActorMap.ActorsInBox(finalTarget, finalSource);
+			var intersectedActors = new List<Actor>();
+
+			foreach (var currActor in actorsInSquare)
+			{
+				long actorWidth;
+				var healthInfo = currActor.Info.TraitInfoOrDefault<HealthInfo>();
+				if (healthInfo != null)
+					actorWidth = healthInfo.Radius.Length;
+				else
+					actorWidth = 0;
+
+				var distance = MinimumDistancePointAndVector(target, source, currActor.CenterPosition);
+				var maxReach = (long)(actorWidth + rayWidth.Length);
+				maxReach *= maxReach;
+
+				// Log.Write("debug", "FindActorsInRaycast, distance : {0}, maxReach : {1}", distance, maxReach);
+				if (distance <= maxReach)
+					intersectedActors.Add(currActor);
+			}
+
+			return intersectedActors;
+		}
+
+		/// <summary>
+		/// Returns the (squared) minimum distance between a line and a point.
+		/// From http://stackoverflow.com/questions/849211/shortest-distance-between-a-point-and-a-line-segment
+		/// </summary>
+		/// <param name="target">The target point (head) of the line</param>
+		/// <param name="source">The source point (tail) of the line</param>
+		/// <param name="point">The point that the minimum distance should be found to</param>
+		/// <returns>The square of the minimum distance between the line and point.</returns>
+		private static long MinimumDistancePointAndVector(WPos target, WPos source, WPos point)
+		{
+			// Return minimum distance between line segment target->source and point
+			var squaredLength = (target - source).LengthSquared;
+
+			// target = source case, so just find distance from the point to the target.
+			if (squaredLength == 0.0) return (target - point).LengthSquared;
+
+			// Consider the line extending the segment, parameterized as target + t (source - target).
+			// We find projection of point onto the line.
+			// It falls where t = [(point - target) . (source - target)] / |source - target|^2
+			long t = DotProduct(target, source, point);
+
+			// Beyond the 'target' end of the segment
+			if (t < 0)
+				return (point - target).LengthSquared;
+
+			// Beyond the 'source end of the segment
+			if (t > squaredLength)
+				return (point - source).LengthSquared;
+
+			// Projection falls on the segment
+			// Using floating points, this would be: var newX = (int)((double)target.X + t * (double)(source.X - target.X));
+			// where t would already be divided by squaredLength.
+			var newX = (int)(target.X + (t * (source.X - target.X) / squaredLength));
+			var newY = (int)(target.Y + (t * (source.Y - target.Y) / squaredLength));
+			var projection = new WPos(newX, newY, (target.Z + source.Z) / 2);
```

Comment:
I need some help on this.
To standardize ```t = 0``` -> ```t = squaredLength``` to ```t = 0``` -> ```t = 1024``` I will have to multiply numerator and then divide by a denominator (if I want to work with purely whole numbers).
I am at a loss as to how to calculate these values without using floating points.

Can I not rather just use ```var projection = WPos.Lerp(source, target, t, squaredLength);``` ? Since the length of the line is ```squaredLength```, and ```t`` is the length along that line at which point we want the new projected position?